File: LoadDLL.c
//Several ways to call functions stored in Dynamic Link Libraries
//Questions? Contact: Leonardo Boselli <boselli@uno.it>

// Only a few functions definitions: No need to load long headers
void printf(const char* fmt, ...);

typedef struct Library Library;
Library* libraryLoad(const char* libName);
void libraryFree(Library* lib);
void* libraryGetProcAddress(Library* lib, const char* funcName);

typedef float (*NormFunc)(float x, float y, float z);
float norm(float x, float y, float z);

// function main() loaded and executed by LUA
int main(void) {
  printf("a) ... the use of libraryGetProcAddress()\n");
  Library* lib = libraryLoad("SampleLibrary.dll");
  NormFunc normFunc = (NormFunc)libraryGetProcAddress(lib,"norm");
  // This way there is no need to use Compiler:addLibrary() from LUA
  float val = normFunc(2,3,4);
  libraryFree(lib);
  printf("The norm of (2,3,4) is %f\n",val);
  //
  printf("b) ... the use of the addLibrary() method\n");
  // The linker looks for norm() in the DLLs added by LUA with Compiler:addLibrary()
  val = norm(2,3,4);
  printf("The norm of (2,3,4) is %f\n",val);
  //
  // Since the function does not return any argument to the LUA caller,
  // it specifies that zero argument were pushed on the LUA stack.
  // That's why this function MUST return zero.
  return 0;
}

File: LoadDLL.lua
----Several ways to call functions stored in Dynamic Link Libraries
----Questions? Contact: leo <tetractys@users.sf.net>

----INIT----
function init()
  showConsole(false)
  ----
  print("\n*** BEGIN ***\n")
  print("\n1) Loading a DLL function through LUA using loadlib()")
  local norm = package.loadlib("SampleLibrary.dll","luaNorm")
  local val = norm(2,3,4)
  print("The norm of (2,3,4) is ",val)
  ----
  print("\n2) Loading a DLL function through C with...")
  cc = Compiler()
  ok = cc:compileFile("LoadDLL.c");
  if ok then
    cc:addLibrary("SampleLibrary.dll")
    ok = cc:link()
  else
    print("\nCompile error")
  end
  if ok then
    c_function = cc:getFunction("main")
  else
    print("\nLink error")
  end
  if c_function then
    c_function()
  else
    print("\nCompiler:getFunction() error")
  end
  print("\n*** END ***\n\n")
  print("Press 'ENTER' to go back to demos menu")
  ----HELP----
  local help = {
    "[ENTER] Demos Menu",
  }
  setHelp(help)
  showHelpReduced()
end

----LOOP----
function update()
end

----FINAL----
function final()
  if cc then
    cc:delete()
    cc = nil
  end
  c_function = nil
  ok = nil
end

----KEYDOWN----
function keyDown(key)
  ----LOAD MAIN MENU----
  if key == string.byte("\r") then
    releaseKey(string.byte("\r"))
    final()
    dofile("main.lua")
    return
  end
end

----SCENE SETUP----
setScene(Scene(init,update,final,keyDown))
